route/build/src/histVars_data.f90

```diff
@@ -260,4 +275,184 @@ SUBROUTINE clean(this)
+      integer(i4b),         intent(out)  :: ierr                  ! error code
+      character(len=strLen),intent(out)  :: message               ! error message
+      ! local variable
+      character(len=strLen)              :: cmessage              ! error message from subroutines
+      real(sp), allocatable              :: array_tmp(:)          ! temp array
+      integer(i4b)                       :: ixRoute, ix1, ix2     ! loop index
+      integer(i4b)                       :: ixFlow, ixVol         ! temporal method index
+      logical(lgt)                       :: FileStatus            ! file open or close
+      integer(i4b), allocatable          :: compdof_rch(:)        ! global reach index for PIO reading
+      integer(i4b), allocatable          :: compdof_hru(:)        ! globa hru index for PIO reading
+      type(file_desc_t)                  :: pioFileDesc           ! pio file handle
+      type(io_desc_t)                    :: ioDescRchHist         ! PIO domain decomposition data for reach flux [nRch]
+      type(io_desc_t)                    :: ioDescHruHist         ! PIO domain decomposition data for hru runoff [nHRU]
+
+      ierr=0; message='read_restart/'
+
+      call openFile(pioSystem, pioFileDesc, trim(restart_name), pio_typename, ncd_nowrite, FileStatus, ierr, cmessage)
+      if(ierr/=0)then; message=trim(message)//trim(cmessage); return; endif
+
+      call get_nc(trim(restart_name), 'history_time', this%timeVar, 1, ierr, cmessage) ! get restart datetime
+      call get_nc(trim(restart_name), 'nt', this%nt, 1, ierr, cmessage)                ! get number of aggretation time steps
+
+      ! get number of HRUs and Reaches for each core
+      if (masterproc) then
+        this%nHru =nHru_mainstem+nHru_trib
+        this%nRch =nRch_mainstem+nRch_trib+nTribOutlet
+        allocate(array_tmp(nRch_mainstem+nRch_trib))
+      else
+        this%nHru = nHru_trib
+        this%nRch = nRch_trib
+        allocate(array_tmp(nRch_trib))
+      end if
+
+      ! global reach and hru index
+      if (masterproc) then
+        allocate(compdof_rch(sum(rch_per_proc(-1:pid))))
+        allocate(compdof_hru(sum(hru_per_proc(-1:pid))))
+      else
+        allocate(compdof_rch(rch_per_proc(pid)))
+        allocate(compdof_hru(hru_per_proc(pid)))
+      end if
+
+      ! For reach flux/volume
+      if (masterproc) then
+        ix1 = 1
+      else
+        ix1 = sum(rch_per_proc(-1:pid-1))+1
+      endif
+      ix2 = sum(rch_per_proc(-1:pid))
+      compdof_rch = arth(ix1, 1, ix2-ix1+1)
+
+      ! For HRU flux/volume
+      if (masterproc) then
+        ix1 = 1
+      else
+        ix1 = sum(hru_per_proc(-1:pid-1))+1
+      endif
+      ix2 = sum(hru_per_proc(-1:pid))
+      compdof_hru = arth(ix1, 1, ix2-ix1+1)
+
+      ! initialze ioDesc
+      call pio_decomp(pioSystem,         & ! input: pio system descriptor
+                      ncd_float,         & ! input: data type (pio_int, pio_real, pio_double, pio_char)
```

Comment:
For restarts you should ALWAYS write out data in double precision. Otherwise restarts won't be exact.

For history files you want to be able to change it in order to save disk space.

Reply:
type defined in pio decomposition needs to be the same as output type (so if we output history variable in single, this has to be ncd_float)

Reply:
changed it to ncd_double and write this in restart file with double precision